bentoml/_internal/context.py

```diff
@@ -1,5 +1,157 @@
 import typing as t
 import contextvars
+from abc import ABC
+from abc import abstractmethod
+from typing import TYPE_CHECKING
+
+if TYPE_CHECKING:
+    import starlette.requests
+    import starlette.responses
+
+
+class Metadata(ABC):
+    @abstractmethod
+    def __setitem__(self, key: str, value: str) -> None:
+        """
+        Set the header ``key`` to ``value``, removing any duplicate entries.
+        Retains insertion order.
+        """
+
+    @abstractmethod
+    def __delitem__(self, key: str) -> None:
+        """
+        Remove the header ``key``.
+        """
+
+    @abstractmethod
+    def __ior__(self, other: t.Mapping[t.Any, t.Any]) -> "Metadata":
+        """
+        Updates this metadata with the contents of ``other``
+        """
+
+    @abstractmethod
+    def __or__(self, other: t.Mapping[t.Any, t.Any]) -> "Metadata":
+        """
+        Returns a new metadata object with the contents of this metadata object updated with the contents of ``other``
+        """
+
+    @abstractmethod
+    def setdefault(self, key: str, value: str) -> str:
+        """
+        If the header ``key`` does not exist, then set it to ``value``.
+        Returns the header value.
+        """
+
+    @abstractmethod
+    def update(self, other: t.Mapping[t.Any, t.Any]) -> None:
+        """
+        Sets all the headers in ``other`` in this object.
+
+        For example, if this object is ``{"my-header": "my-value", "my-other-header": "my-other-value"}``
+        and other is {"my-header": 3¸ "other-header": 4}
+        """
+
+    @abstractmethod
+    def append(self, key: str, value: str) -> None:
+        """
+        Append a header, preserving any duplicate entries.
+        """
+
+
+class InferenceApiContext:
```

Comment:
`set_headers` missing?

Reply:
I don't think we intend to have a `set_headers` method.